app/modules/pwm2.c

```diff
@@ -0,0 +1,553 @@
+  ets_printf("pwm2 : interruptTimerTicks %u\n", moduleData->setupData.interruptTimerTicks);
+  ets_printf("pin, duty, pulse, resolutionAsCPU, resolutionIntrCntMultiplier, divisableFreq, freqDivisor\n");
+  for (int i = 1; i < GPIO_PIN_NUM; i++)
+  {
+    if (isPinSetup(moduleData, i))
+    {
+      ets_printf("%u, %u, %u, %u, %u, %u, %u\n",
+                 i,
+                 moduleData->setupData.pin[i].duty,
+                 moduleData->setupData.pin[i].pulseResolutions,
+                 moduleData->setupData.pin[i].resolutionCPUTicks,
+                 moduleData->setupData.pin[i].resolutionInterruptCounterMultiplier,
+                 moduleData->setupData.pin[i].divisableFrequency,
+                 moduleData->setupData.pin[i].frequencyDivisor);
+    }
+  }
+  ets_printf("pwm2 : enabledGpioMask %X\n", moduleData->interruptData.enabledGpioMask);
+  ets_printf("pin, pulseAsInterruptCnt, pinOffWhenIntrCnt, currentintrCnt, gpioMask\n");
+  for (int i = 1; i < GPIO_PIN_NUM; i++)
+  {
+    if (isPinSetup(moduleData, i))
+    {
+      ets_printf("%u, %u, %u, %u, %X\n",
+                 i,
+                 moduleData->interruptData.pin[i].pulseInterruptCcounter,
+                 moduleData->interruptData.pin[i].offInterruptCounter,
+                 moduleData->interruptData.pin[i].currentInterruptCounter,
+                 moduleData->interruptData.pin[i].gpioMask);
+    }
+  }
+}
+
+static uint64_t enduserFreqToCPUTicks(const uint64_t divisableFreq, const uint64_t freqDivisor, const uint64_t resolution)
+{
+  return (getCPUTicksPerSec() / (freqDivisor * resolution)) * divisableFreq;
+}
+
+static uint16_t getPinGpioMask(uint8_t pin)
+{
+  return 1 << GPIO_ID_PIN(pin_num[pin]);
+}
+
+static void set_duty(pwm2_module_data_t *moduleData, const uint8_t pin, const uint32_t duty)
+{
+  pwm2_pin_setup_t *sPin = &moduleData->setupData.pin[pin];
+  pwm2_pin_interrupt_t *iPin = &moduleData->interruptData.pin[pin];
+  sPin->duty = duty;
+  updatePinOffCounter(iPin, sPin);
+}
+
+static int setup_pin(
+    pwm2_module_data_t *moduleData,
+    const uint8_t pin,
+    const uint32_t divisableFreq,
+    const uint32_t freqDivisor,
+    const uint32_t resolution,
+    const uint32_t initDuty)
+{
+  moduleData->setupData.pin[pin].pulseResolutions = resolution;
+  moduleData->setupData.pin[pin].divisableFrequency = divisableFreq;
+  moduleData->setupData.pin[pin].frequencyDivisor = freqDivisor;
+  moduleData->setupData.pin[pin].resolutionCPUTicks = enduserFreqToCPUTicks(divisableFreq, freqDivisor, resolution);
+  moduleData->interruptData.pin[pin].gpioMask = getPinGpioMask(pin);
+  reCalculateCommonToAllPinsData(moduleData);
+  set_duty(moduleData, pin, initDuty);
+}
+
+static int lpwm2_setup_pin_common(lua_State *L, const bool isFreqHz)
+{
+  if (moduleData->setupData.isStarted)
+  {
+    return luaL_error(L, "pwm2 : already started, stop it before setting up pins.");
+  }
+  int pos = 0;
+  while (true)
```

Comment:
This code does not match the documentation for `pwm2.setup_pin_hz(pin,frequencyAsHz,pulsePeriod,initialDuty [,frequencyDivisor])` (and ditto the sec variant).  Eg. why the `while` loop and why are these mandatory parameters optional?  I would have expected something more like:
```C
  int pin           = lua_tointeger(L, 1);
  int freq          = lua_tointeger(L, 2);
  int resolution    = lua_tointeger(L, 3);
  int freqFractions = lua_tointeger(L, 4);
  int initDuty      = lua_tointeger(L, 5);
  
  luaL_argcheck(L, pin > 0 && pin <= GPIO_PIN_NUM, 1, "invalid pin number");
  luaL_argcheck(L, freq > 0, 2, "invalid frequency");
  luaL_argcheck(L, resolution > 0, 2, "invalid resolution");
  // ...
```
or whatever is consistent with the documentation.  The Lua RTS rolls up error reporting as standard.  There's little to be gained by doubling up on this.

Reply:
nice spotting, the loop was from a past idea not relevant anymore. and thanks for the examples, now should be ok.